### PR TITLE
Switch from std::regex to boost::regex

### PR DESCRIPTION
**Pull request by yorickvP**
# Motivation
Requested in #7336

# Context
Fixes https://github.com/NixOS/nix/issues/2147, fixes https://github.com/NixOS/nix/issues/4758
See also https://github.com/NixOS/nix/pull/3826.

- `s/std::regex/regex/g`, added aliases to enable switching back and forth in the future.
- Disabled enableIcu on boost dep, but this hasn't taken effect yet because this change isn't in nixpkgs 22.11.

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or bug fix: updated release notes

### CODE REVIEW

**Comment by edolstra** [comment]
I guess we should merge this after the `enableIcu` change has landed in stable Nixpkgs.

File: flake.nix [hunk 1]
```diff
@@ -701,7 +711,7 @@
 
             outputs = [ "out" "dev" "doc" ];
 
-            nativeBuildInputs = nativeBuildDeps;
+            nativeBuildInputs = nativeBuildDeps ++ [ clang-tools ];
```
Comment (thufschmitt) [inline flake.nix:714]:
```suggestion
            nativeBuildInputs = nativeBuildDeps;
```

I gues this sneaked in accidentally

**Review by thufschmitt — COMMENTED** [review]
Nice :)

Can you confirm that the closure size doesn't increase too much (with a nixpkgs that has the `enableIcu` patch)?

Other than that, I agree with Eelco, let's just wait for `enableIcu` to land in nixpkgs (can you link to the PR for that?), and then we can merge

**Comment by yorickvP** [comment]
enableIcu is https://github.com/NixOS/nixpkgs/pull/205166, but not backported to 22.11

Sizes:
```
$ nix path-info ./master
/nix/store/dlzwlll6d5k5yjfg98q51szgzxgqhmf7-nix-2.14.0pre20230203_895dfc6                  	  16.7M	  88.1M

$ nix path-info ./boost-regex
/nix/store/r3x96j3kmcs8dv4l02rrjmbhm535jycy-icu4c-72.1                                     	  36.8M	  76.7M
/nix/store/xyxm8cp19vrqnfssvhkm6k50ry0gaqhj-nix-2.14.0pre20230207_dirty                    	  17.3M	 125.5M

$ nix path-info ./boost-regex-without-icu
/nix/store/74jhsx372zq72hga6l2llfks700p3bam-nix-2.14.0pre20230207_dirty                    	  17.3M	  88.7M
```

**Comment by thufschmitt** [comment]
(answering https://github.com/NixOS/nix/pull/7336#issuecomment-1422392607 here since it's where the `boost` implementation takes place):

@SuperSandro2000 that's a very valid point. @yorickvP do you think you could benchmark this quickly to see whether it has an impact? Something like `hyperfine --warmup 1 "nix build nix/895dfc656a21f6252ddf48df0d1f215effa04ecb --rebuild" "nix build nix/d06b144dc38f509bb35f940b38d810a94ad5b852` on a quiet machine should do it

**Comment by thufschmitt** [comment]
I think a 2s difference is quite OK. I benchmarked the time of a full `nix build` (it's cold out there, needed to heat my office a bit), and the difference is neglegible

```
Summary
  'nix build nix/d06b144dc38f509bb35f940b38d810a94ad5b852 --rebuild' ran
    1.00 ± 0.01 times faster than 'nix build nix/895dfc656a21f6252ddf48df0d1f215effa04ecb --rebuild'
```

**Comment by SuperSandro2000** [comment]
The difference should be much bigger when boost would be fully removed which is a design decision completely out of scope of this PR.

See https://www.factorio.com/blog/post/fff-206#:~:text=step%203%20-%20getting%20rid%20of%20boost for an example and some stats

**Comment by yorickvP** [comment]
@SuperSandro2000 Factorio looks like it was using a lot more boost than Nix.

it might be worth it to remove boost::lexical_cast from util.hh or boost::format from logging, but the rest of the boost usage seems to be very contained and not likely to matter much for compile times.

I don't predict more than a 10% compile time increase if we replace it.

**Comment by yorickvP** [comment]
Nix team meeting notes:
> [@regnat](https://discourse.nixos.org/u/regnat) had talked to [@yorickvP](https://discourse.nixos.org/u/yorickvp), will continue after next NixOS stable so we can use for next Nix release.

I think we should merge it so that it ends up in 23.05, as well. Would that be possible?

**Comment by con-f-use** [comment]
Any reason why this never got merged?

**Comment by yorickvP** [comment]
Needed to wait for enableIcu to land on the boost dependency, which I think is the case now.

**Comment by fricklerhandwerk** [comment]
@yorickvP since the preconditions are met, can you rebase?

**Comment by iFreilicht** [comment]
As far as I understand, this would also fix #8768 (at least the inconsistency part), #3047 and #1537.

**Comment by yorickvP** [comment]
Rebased!

File: flake.nix [hunk 2]
```diff
@@ -495,6 +499,12 @@
             meta.mainProgram = "nix";
           });
 
+          boost-nix = final.boost.override {
+            # enableIcu arg is not yet supported
+            # but will be with next nixpkgs update
```
Comment (thufschmitt) [inline flake.nix:504]:
```suggestion
```

**Comment by thufschmitt** [comment]
Thanks, @yorickvP !

Set to auto-merge since it was already approved on principle.

**Comment by 9999years** [comment]
This broke `lib.escapeRegex` from nixpkgs. Before:

```
$ nix repl nixpkgs
Welcome to Nix 2.18.1. Type :? for help.

Loading installable 'flake:nixpkgs#'...
Added 5 variables.
nix-repl> builtins.match (lib.escapeRegex "{}") "{}"
[ ]
```

After:

```
$ ~/nix/result/bin/nix repl nixpkgs
Welcome to Nix 2.20.0pre20231130_dirty. Type :? for help.

Loading installable 'flake:nixpkgs#'...
Added 5 variables.
nix-repl> builtins.match (lib.escapeRegex "{}") "{}"
error:
       … while calling the 'match' builtin

         at «string»:1:1:

            1| builtins.match (lib.escapeRegex "{}") "{}"
             | ^

       error: invalid regular expression '\{}'
```

**Comment by infinisil** [comment]
This should be reverted, I opened https://github.com/NixOS/nix/pull/9508 to do that

**Comment by thufschmitt** [comment]
Thanks for the revert @infinisil !

Could have been caught earlier?. Is there a nixpkgs lib test suite that we can use? (We already test that the output  of `nix-env -qaP --drv-path -f <someFixedNixpkgsVersion>` doesn't change, but that's apparently not enough)

**Comment by roberth** [comment]
We have `nixpkgsLibTests` on Hydra. It runs `${nixpkgs}/lib/tests/release.nix` with the fresh Nix.

**Comment by yorickvP** [comment]
Hrm, sad. I hope it just needs `boost::regex::extended`. Should test.

**Comment by roberth** [comment]
`lib` is only responsible for testing itself, but we use it here as a cheap way to hopefully extend our own coverage a bit.

**Pull request merged**